scripts/e2e-host-controller-stack.sh

```
@@ -0,0 +1,379 @@
+prepare_stack_helpers() {
+  local project="$1" source_dir="$2" helpers="$3" run_id="$4" attempt="$5" phase="$6" scenario="$7"
+  DENSE_MEM_CI_HELPER_DIR="${RUN_DIR}/${run_id}-${attempt}/${phase}-${scenario}-helpers"
+  DENSE_MEM_CI_PRIVATE_DIR="${RUN_DIR}/${run_id}-${attempt}/${phase}-${scenario}-private"
+  DENSE_MEM_CI_COMPOSE_OVERLAY_FILE=""
+  mkdir -p "$DENSE_MEM_CI_HELPER_DIR" "$DENSE_MEM_CI_PRIVATE_DIR"
+  chmod 700 "$DENSE_MEM_CI_HELPER_DIR" "$DENSE_MEM_CI_PRIVATE_DIR"
+  [[ -d "$source_dir" && "$source_dir" == /* ]] || fail "helper source directory must be an absolute directory"
+
+  local oauth_token=""
+  local harness_image=""
+  local provider_dimensions="$(env_value AI_API_EMBEDDING_DIMENSIONS 2>/dev/null || printf '%s' 1536)"
```

Shellcheck (0.11.0)

[warning] Declare and assign separately to avoid masking return values.

(SC2155)